Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,164 @@
+/-
+Copyright (c) 2024 Elliot Dean Young and Jiazhen Xia. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiazhen Xia, Elliot Dean Young
+-/
+
+import Mathlib.Topology.Category.TopCat.Limits.Basic
+import Mathlib.Analysis.InnerProductSpace.PiL2
+
+/-!
+# CW-complexes
+
+In this file we define a relative CW-complex as the colimit of an expanding sequence of subspaces
+`sk i` (called the `i`-skeleta) for `i ≥ -1`, where `sk (-1)` is an arbitrary topological space
+and each `sk (n+1)` is obtained from `sk n` by attaching (n+1)-disks.
+
+## References
+The definition of CW complexes follows David Wärn's suggestion at
+https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Do.20we.20have.20CW.20complexes.3F/near/231769080
+-/
+
+open CategoryTheory
+
+namespace CWComplex
+
+/-- The n-dimensional sphere is the set of points in ℝ^(n+1) whose norm equals 1, endowed with the
+subspace topology. -/
+noncomputable def sphere (n : ℤ) : TopCat :=
```

Comment:
This must be `TopCat.{0}`. I would be interested to know how difficult it would be to generalize definitions to `TopCat.{u}` so that in a certain future, we could relate simplicial sets in `Type u` with CW-complexes.{u}. (I do not want to push too much in this direction if it is painful to implement.)

Another (more important) suggestion is that with this definition of the basic cells, it may not be very easy to show that a sphere or a ball is a CW-complex. Topologically, as a substitute, we could use the topological realizations of the simplicial set inclusions `∂Δ[n] ⟶ Δ[n]` (these are isomorphic arrows in the category of topological spaces). Then, it would be much easier to show that the realization of simplicial sets are CW-complexes using the skeleton filtration (which would require a little work from the notions introduced in `AlgebraicTopology.SplitSimplicialObject`).

Then, in any case, I think that `AttachCells` should be generalized as `AttachGeneralizedCells` which would take as input `X`, `X'` and a continuous map of topological spaces (typically your `sphereInclusion`, but also the isomorphic arrows I am suggesting using simplicial sets).

Reply:
Hi Joël.

Thanks very much for your time. I'd like to say I'm especially excited about the final suggestion you made, but first I want to explain how a sphere has a CW-approximation in this implementation:

If a CW-approximation exists for a space X homeomorphic to a space Y, then one exists for Y. Meanwhile the cube has a CW-approximation, merely one which features only one cell. Of course, one homeomorphism from the ∞-norm unit ball in ℝⁿ to the 2-norm unit ball in ℝⁿ is division by the 2-norm.

Despite the above, I'm really excited about your idea to include all of the basic shapes which make spheres and discs:

- the unit ball in the 2 norm its boundary
- the unit ball in the ∞ norm and its boundary
- the simplex and its boundary

I think we could start by defining a CW-complex as resulting from any of the three, but also to include homeomorphisms between Δⁿ, Dⁿ, and Iⁿ which restrict homeomorphisms of ∂Δⁿ, ∂Dⁿ, and ∂Iⁿ.

One further question is whether to include the p-norm unit ball for 1 ≤ p ≤ ∞.

Do you think that including all three kinds of attaching maps along with the mentioned homeomorphisms would satisfy the last issue?

One more thing I didn't mention is that CW-replacement and transport of πₙ (base-replacement) remains to be shown, but we thought it would be best to make this first smaller PR.

Reply:
If I want to show that topological realizations `|X|` of simplicial sets are CW-complexes, I would inductively attach a `n`-cell for each nondegenerate `n`-simplex. For example, for `|Δ[2]|`, this would mean starting with the empty set, add 3 points `0`, `1`, `2`, add 3 segments `[0,1]`, `[1,2]`, `[0,2]` to get a triangle (without its interior), and then attach a `2`-disk. This way, we get that the `2`-disk is a CW-complex.

Then, `Δ[n]/∂Δ[n]` has only two nondegenerate simplices, one `0`-simplex and one `n`-simplex. In this case attaching the `n` cell means identifying all the boundary points (which are on a sphere) of the unit ball, which gives a space that is homeomorphic (but not so obviously) to the n-dimensional sphere.

I do not think that the unit ball can be obtained using only one (or maybe two) cells. For the usual disk above, I have used `7` cells!

My more precise suggestion is to define a type `AttachGeneralizedCells X X' f` (where `f : A → B` gives the sort of "generalized" cells) which contains the data which says that `X'` is obtained from `X` by attaching `f`-cells. Then, if `f` and `g` are isomorphic arrows in `TopCat`, construct a bijection between `AttachGeneralizedCells X X' f` and `AttachGeneralizedCells X X' f'`. Then `AttachCells X X' n` can be defined as an abbreviation of this for the inclusion of the boundary of the ball. Then, the definition of relative CW-complexes is unchanged, but for their construction, we may now use more convenient types of cells using the bijections above. I do not think that more explicit types of cells are necessary besides the topological sphere/ball and the inclusions `|∂Δ[n]| → |Δ[n]|`.

Reply:
Hi Joël,

Thank you for your review. I have updated my code according to your suggestions.

> I would be interested to know how difficult it would be to generalize definitions to `TopCat.{u}` so that in a certain future, we could relate simplicial sets in `Type u` with CW-complexes.{u}.

All definitions are now universe polymorphic. This was not difficult but I was unaware of its benefits, so thanks for reminding me.

> My more precise suggestion is to define a type `AttachGeneralizedCells X X' f`... Then `AttachCells X X' n` can be defined as an abbreviation

https://github.com/leanprover-community/mathlib4/blob/3cfda01c25cc8ef045b74478e0ccc1f6dbb3571c/Mathlib/Topology/CWComplex.lean#L60

https://github.com/leanprover-community/mathlib4/blob/3cfda01c25cc8ef045b74478e0ccc1f6dbb3571c/Mathlib/Topology/CWComplex.lean#L77-L79

https://github.com/leanprover-community/mathlib4/blob/3cfda01c25cc8ef045b74478e0ccc1f6dbb3571c/Mathlib/Topology/CWComplex.lean#L96-L97